// rettiwt-api/dist/enums/ApiErrors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EApiErrors = void 0;
/**
 * The different types of api error messages.
 *
 * @public
 */
var EApiErrors;
(function (EApiErrors) {
    EApiErrors["COULD_NOT_AUTHENTICATE"] = "Failed to authenticate";
    EApiErrors["RESOURCE_NOT_FOUND"] = "Requested resource not found";
    EApiErrors["MISSING_PARAMETER"] = "Missing named parameter";
    EApiErrors["USER_NOT_FOUND"] = "Requested user not found";
    EApiErrors["USER_SUSPENDED"] = "Requested user has been suspended";
    EApiErrors["ACCOUNT_SUSPENDED"] = "Account has been suspended";
    EApiErrors["RATE_LIMIT_EXCEEDED"] = "Rate limit exceeded";
    EApiErrors["INTERNAL_ERROR"] = "Internal server error";
    EApiErrors["TIME_ERROR"] = "Mismatched data/time with server";
    EApiErrors["ALREADY_FAVORITED"] = "Tweet already favorited";
    EApiErrors["STATUS_NOT_FOUND"] = "Requested tweeet not found";
    EApiErrors["NOT_AUTHORIZED"] = "Not authorized to view tweet";
    EApiErrors["DAILY_STATUS_LIMIT_EXCEEDED"] = "Exceeded daily tweet update limit";
    EApiErrors["TWEET_LENGTH_EXCEEDED"] = "Exceeded tweet text maximum length";
    EApiErrors["DUPLICATE_STATUS"] = "Tweet already posted";
    EApiErrors["BAD_AUTHENTICATION"] = "Invalid authentication data";
    EApiErrors["RESOURCE_NOT_ALLOWED"] = "Not authorized to access requested resource";
    EApiErrors["AUTOMATED_REQUEST_ERROR"] = "Automated request detected";
    EApiErrors["ACCOUNT_LOCKED"] = "Account has been locked";
    EApiErrors["ALREADY_RETWEETED"] = "Tweet already retweeted";
    EApiErrors["TWEET_NOT_FOUND"] = "Requested tweet not found";
    EApiErrors["TWEET_VIOLATED_RULES"] = "Requestd tweet has been removed for rules violation";
    EApiErrors["DISABLED_TWEET_ACTIONS"] = "Reqeusted action disabled on the tweet";
})(EApiErrors || (exports.EApiErrors = EApiErrors = {}));
