// rettiwt-core/dist/models/Request.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Request = void 0;
// ENUMS
var Request_1 = require("../enums/Request");
var Resources_1 = require("../enums/Resources");
// MODELS
var Query_1 = require("./params/Query");
/**
 * The request containing all the required url, params, query, payload, etc for a requested resource on Twitter.
 *
 * @public
 */
var Request = /** @class */ (function () {
    /**
     * Generates an HTTP request configuration for the requested resource on Twitter.
     *
     * @param resourceType - The type of resource requested.
     * @param args - Additional URL arguments.
     */
    function Request(resourceType, args) {
        /** The base URL of the request. */
        this.base = 'https://twitter.com';
        this.url = "".concat(this.base).concat(resourceType);
        this.endpoint = resourceType;
        this.payload = new Query_1.Query(resourceType, args);
        // For 'POST' requests
        if (resourceType == Resources_1.EResourceType.CREATE_TWEET ||
            resourceType == Resources_1.EResourceType.CREATE_RETWEET ||
            resourceType == Resources_1.EResourceType.FAVORITE_TWEET) {
            this.type = Request_1.ERequestType.POST;
        }
        // For 'GET' requests
        else {
            this.type = Request_1.ERequestType.GET;
            this.url = "".concat(this.url, "?").concat(this.payload.toString());
        }
    }
    return Request;
}());
exports.Request = Request;
