// rettiwt-core/dist/models/payloads/Features.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Features = void 0;
/**
 * Parameters for customizing the raw response, that must be sent as a URL-encoded, stringified-JSON.
 *
 * @public
 */
var Features = /** @class */ (function () {
    function Features() {
        /* eslint-disable @typescript-eslint/naming-convention */
        this.rweb_lists_timeline_redesign_enabled = true;
        this.responsive_web_graphql_exclude_directive_enabled = true;
        this.verified_phone_label_enabled = true;
        this.creator_subscriptions_tweet_preview_api_enabled = true;
        this.responsive_web_graphql_timeline_navigation_enabled = true;
        this.responsive_web_graphql_skip_user_profile_image_extensions_enabled = false;
        this.responsive_web_home_pinned_timelines_enabled = false;
        this.tweetypie_unmention_optimization_enabled = true;
        this.responsive_web_edit_tweet_api_enabled = true;
        this.graphql_is_translatable_rweb_tweet_is_translatable_enabled = true;
        this.view_counts_everywhere_api_enabled = true;
        this.longform_notetweets_consumption_enabled = true;
        this.responsive_web_twitter_article_tweet_consumption_enabled = false;
        this.tweet_awards_web_tipping_enabled = false;
        this.freedom_of_speech_not_reach_fetch_enabled = true;
        this.standardized_nudges_misinfo = true;
        this.tweet_with_visibility_results_prefer_gql_limited_actions_policy_enabled = true;
        this.longform_notetweets_rich_text_read_enabled = true;
        this.longform_notetweets_inline_media_enabled = true;
        this.responsive_web_media_download_video_enabled = false;
        this.responsive_web_enhance_cards_enabled = false;
        this.hidden_profile_likes_enabled = false;
        this.hidden_profile_subscriptions_enabled = false;
        this.subscriptions_verification_info_verified_since_enabled = true;
        this.highlights_tweets_tab_ui_enabled = true;
    }
    /* eslint-enable @typescript-eslint/naming-convention */
    /**
     * Converts this object to it's string representation.
     *
     * @returns 'this' object's string representation.
     */
    Features.prototype.toString = function () {
        return "".concat(encodeURIComponent(JSON.stringify(this)));
    };
    return Features;
}());
exports.Features = Features;
