// rettiwt-api/dist/models/public/User.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.User = void 0;
/**
 * The details of a single user.
 *
 * @public
 */
var User = /** @class */ (function () {
    /**
     * Initializes a new User from the given raw user data.
     *
     * @param user - The raw user data.
     */
    function User(user) {
        this.id = user.rest_id;
        this.userName = user.legacy.screen_name;
        this.fullName = user.legacy.name;
        this.createdAt = user.legacy.created_at;
        this.description = user.legacy.description;
        this.isVerified = user.legacy.verified;
        this.favouritesCount = user.legacy.favourites_count;
        this.followersCount = user.legacy.followers_count;
        this.followingsCount = user.legacy.friends_count;
        this.statusesCount = user.legacy.statuses_count;
        this.location = user.legacy.location;
        this.pinnedTweet = user.legacy.pinned_tweet_ids_str[0];
        this.profileBanner = user.legacy.profile_banner_url;
        this.profileImage = user.legacy.profile_image_url_https;
    }
    return User;
}());
exports.User = User;
