// rettiwt-auth/dist/enums/Login.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ELoginSubtasks = exports.ELoginUrls = void 0;
/**
 * The various urls for various operations related to login.
 *
 * @internal
 */
var ELoginUrls;
(function (ELoginUrls) {
    ELoginUrls["GUEST_TOKEN"] = "https://api.twitter.com/1.1/guest/activate.json";
    ELoginUrls["INITIATE_LOGIN"] = "https://api.twitter.com/1.1/onboarding/task.json?flow_name=login";
    ELoginUrls["LOGIN_SUBTASK"] = "https://api.twitter.com/1.1/onboarding/task.json";
})(ELoginUrls = exports.ELoginUrls || (exports.ELoginUrls = {}));
/**
 * The various subtasks involved in logging in to a Twitter account.
 *
 * @internal
 */
var ELoginSubtasks;
(function (ELoginSubtasks) {
    ELoginSubtasks["JS_INSTRUMENTATION"] = "LoginJsInstrumentationSubtask";
    ELoginSubtasks["ENTER_USER_IDENTIFIER"] = "LoginEnterUserIdentifierSSO";
    ELoginSubtasks["ENTER_ALTERNATE_USER_IDENTIFIER"] = "LoginEnterAlternateIdentifierSubtask";
    ELoginSubtasks["ENTER_PASSWORD"] = "LoginEnterPassword";
    ELoginSubtasks["ACCOUNT_DUPLICATION_CHECK"] = "AccountDuplicationCheck";
})(ELoginSubtasks = exports.ELoginSubtasks || (exports.ELoginSubtasks = {}));
