// rettiwt-core/dist/enums/Resources.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EResourceType = void 0;
/**
 * The different types of resources that can be fetched.
 *
 * @public
 */
var EResourceType;
(function (EResourceType) {
    EResourceType["CREATE_RETWEET"] = "/i/api/graphql/ojPdsZsimiJrUGLR1sjUtA/CreateRetweet";
    EResourceType["CREATE_TWEET"] = "/i/api/graphql/tTsjMKyhajZvK4q76mpIBg/CreateTweet";
    EResourceType["FAVORITE_TWEET"] = "/i/api/graphql/lI07N6Otwv1PhnEgXILM7A/FavoriteTweet";
    EResourceType["LIST_DETAILS"] = "/i/api/graphql/gO1_eYPohKYHwCG2m-1ZnQ/ListByRestId";
    EResourceType["LIST_TWEETS"] = "/i/api/graphql/naCjgapXCSCsbZ7qnnItQA/ListLatestTweetsTimeline";
    EResourceType["TWEET_SEARCH"] = "/i/api/graphql/nK1dw4oV3k4w5TdtcAdSww/SearchTimeline";
    EResourceType["TWEET_DETAILS"] = "/i/api/graphql/0hWvDhmW8YQ-S_ib3azIrw/TweetResultByRestId";
    EResourceType["TWEET_FAVORITERS"] = "/i/api/graphql/9XKD3EWWC2BKpIFyDj4KKQ/Favoriters";
    EResourceType["TWEET_RETWEETERS"] = "/i/api/graphql/v5h-KLmyl-wqZ8i-a_q73w/Retweeters";
    EResourceType["USER_DETAILS"] = "/i/api/graphql/oUZZZ8Oddwxs8Cd3iW3UEA/UserByScreenName";
    EResourceType["USER_DETAILS_BY_ID"] = "/i/api/graphql/i_0UQ54YrCyqLUvgGzXygA/UserByRestId";
    EResourceType["USER_FOLLOWING"] = "/i/api/graphql/V8zHw0SZijWORSsb-FNrng/Following";
    EResourceType["USER_FOLLOWERS"] = "/i/api/graphql/6y5TB_HrwQM0FBGDiNfoEA/Followers";
    EResourceType["USER_LIKES"] = "/i/api/graphql/kgZtsNyE46T3JaEf2nF9vw/Likes";
    EResourceType["USER_TWEETS"] = "/i/api/graphql/H8OOoI-5ZE4NxgRr8lfyWg/UserTweets";
    EResourceType["USER_TWEETS_AND_REPLIES"] = "/i/api/graphql/GJuqLMoVxFZOvy8Oqec_9Q/UserTweetsAndReplies";
})(EResourceType = exports.EResourceType || (exports.EResourceType = {}));
